src/Orleans.Core/Streams/QueueAdapters/BatchContainerBatch.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// A batch of batch containers, that if configured (see StreamPullingAgentOptions), will be the data pulled by the
+    /// PersistenStreamPullingAgent from it's underlying cache
+    /// </summary>
+    class BatchContainerBatch : IBatchContainerBatch
+    {
+        /// <summary>
+        /// Stream identifier for the stream this batch is part of.
+        /// Derived from the first batch container in the batch.
+        /// </summary>
+        public Guid StreamGuid { get; }
+
+        /// <summary>
+        /// Stream Sequence Token for the start of this batch.
+        /// Derived from the first batch container in the batch.
+        /// </summary>
+        public StreamSequenceToken SequenceToken { get; }
+
+        /// <summary>
+        /// Stream namespace for the stream this batch is part of.
+        /// Derived from the first batch container in the batch.
+        /// </summary>
+        public string StreamNamespace { get; }
+
+        /// <summary>
+        /// Batch containers comprising this batch
+        /// </summary>
+        public IEnumerable<IBatchContainer> BatchContainers { get; }
+
+        public BatchContainerBatch(IEnumerable<IBatchContainer> batchContainers)
+        {
+            if ((batchContainers == null) || !batchContainers.Any())
```

Comment:
Using IEnumerable<T> in this way can allow for multiple enumerations.  Suggest either converting it to a List<T> or using List<T> as input arg